### PR TITLE
Added Packages, Controllers are done.

### PR DESCRIPTION
**Pull request by ismaildev7**
### Remaining Tasks

- Config File
- Error Handler
- Deployment + MongoDB Atlas

### CODE REVIEW

File: controllers/patientController.js [hunk 1]
```diff
@@ -1,183 +1,144 @@
...
 // @route PUT /api/patients/:id
 // @acess Public
-const updatePatient = async (req, res) => {
-    const {error} = validate(req.body)
-    if(error) {
-        // 400 - Bad Request
-        return res.status(400).send(error.details[0].message)
+const updatePatient = asyncHandler(async (req, res) => {
+  const { error } = validate(req.body);
+  if (error) {
```
Comment (bilalashraf1710) [inline controllers/patientController.js:65]: We can improve it, like em seeing the same code in each endpoint so may be a function that validates the body

File: controllers/patientController.js [hunk 2]
```diff
@@ -1,183 +1,144 @@
...
-            acc.push(petDetail)
-            return acc
-        }, [])
-
-        res.send(result)
-    } catch (err) {
-        console.log("Error: ", err.message)
-        res.status(404).send(err.message)
+const getMostPopularPet = asyncHandler(async (req, res) => {
+  const patients = await Patient.find().populate("appointments", "fee isPaid");
```
Comment (bilalashraf1710) [inline controllers/patientController.js:96]: Fee isPaid is an enum? if yes then just use paid/unpaid

File: controllers/patientController.js [hunk 3]
```diff
@@ -1,183 +1,144 @@
...
+
+  const result = {};
+
+  let max = 0;
+  result.popularPet = patients.reduce((acc, patient) => {
```
Comment (bilalashraf1710) [inline controllers/patientController.js]: Can be divide into function so if we need popular pet in some other function then we can just call to that function

File: controllers/patientController.js [hunk 4]
```diff
@@ -1,183 +1,144 @@
...
 
 // Exports Here
-module.exports.createPatient = createPatient
-module.exports.getPatients = getPatients
-module.exports.getPatientById = getPatientById
-module.exports.updatePatient = updatePatient
-module.exports.deletePatient = deletePatient
-module.exports.getMostPopularPet = getMostPopularPet
\ No newline at end of file
+module.exports.createPatient = createPatient;
```
Comment (bilalashraf1710) [inline controllers/patientController.js:106]: module.exports = {
createPatien: cretaPatient,
getPatient... 
}???
POSSIBLE?

File: index.js [hunk 5]
```diff
@@ -1,28 +1,37 @@
 const morgan = require('morgan')
 const Joi = require('joi')
 Joi.objectId = require('joi-objectid')(Joi)
-const mongoose = require('mongoose')
 const express = require('express')
+const colors = require("colors")
+const dotenv = require("dotenv").config()
```
Comment (bilalashraf1710) [inline index.js:6]: yup we should have an idea about dotenv.. but can we use .env ???

Comment (ismaildev7) [reply index.js:6]: yes, we can use it. Basically, I have created a .env file in the project. This .env file push environment variables to the process. Then, to use that .env file, we use dotenv package. This package loads environment variables from a . env file into the running process.

File: controllers/patientController.js (hunk 4 above)
Comment (ismaildev7) [reply controllers/patientController.js:106]: yes, It is possible. **I have used this method before** But in the mosh course, mosh exported the functions by this method. So, I followed mosh method in this Project.

File: controllers/patientController.js (hunk 3 above)
Comment (ismaildev7) [reply controllers/patientController.js]: yep, I will make this function, and also I will try to divide any other functions if I thought They should be divided.

File: controllers/patientController.js (hunk 2 above)
Comment (ismaildev7) [reply controllers/patientController.js:96]: it's not an enum. It's a boolean. By default, isPaid is false meaning the Patient has not paid the fee. we will change isPaid to true when the Patient will pay the fee. Also, I am using a currency field which is an enum like USD, Eur, and Bitcoin.

File: controllers/patientController.js (hunk 1 above)
Comment (ismaildev7) [reply controllers/patientController.js:65]: I already have a function name **validate** which validates the body. I have made this function in the model file.

File: controllers/patientController.js [hunk 6]
```diff
@@ -1,183 +1,116 @@
...
-    } catch (err) {
-        console.log("Error: ", err.message)
-        res.status(404).send(err.message)
-    }
-}
+const getPatientById = asyncHandler(async (req, res) => {
+  const patient = await Patient.findById(req.params.id);
+  if (!patient) {
+    res.status(404);
+    throw new Error("The patient with given ID was not found");
```
Comment (Shazil2154) [inline controllers/patientController.js]: `The patient with Id ${req.params.id} was not found`

File: controllers/patientController.js [hunk 7]
```diff
@@ -1,183 +1,116 @@
...
-    }
-}
+const createPatient = asyncHandler(async (req, res) => {
+  const { error } = validate(req.body);
+  if (error) {
+    res.status(400);
+    throw new Error(error.details[0].message);
+  }
+
+  const newPatient = new Patient({
```
Comment (Shazil2154) [inline controllers/patientController.js:48]: const {name, type, ownerName, ownerAddress, ownerPhoneNo} = req.body;

File: controllers/patientController.js [hunk 8]
```diff
@@ -1,183 +1,116 @@
...
+
+  await newPatient.save();
+  res.status(201).send(newPatient);
+});
 
 // @desc Update a patient
 // @route PUT /api/patients/:id
 // @acess Public
-const updatePatient = async (req, res) => {
-    const {error} = validate(req.body)
```
Comment (Shazil2154) [inline controllers/patientController.js:78]: you can create a middleware for validation

File: controllers/patientController.js [hunk 9]
```diff
@@ -1,183 +1,116 @@
...
-        return res.status(400).send(error.details[0].message)
+const updatePatient = asyncHandler(async (req, res) => {
+  const { error } = validate(req.body);
+  if (error) {
+    res.status(400);
+    throw new Error(error.details[0].message);
+  }
+
+  let patient = await Patient.findByIdAndUpdate(
+    req.params.id,
```
Comment (Shazil2154) [inline controllers/patientController.js:71]: same as above

File: index.js [hunk 10]
```diff
@@ -1,28 +1,37 @@
...
+connectDB()
 
-mongoose.connect('mongodb://localhost/houseTable')
-    .then(() => console.log('Connected to MongoDb...'))
-    .catch((err) => console.log('Could not connect to MongoDb...', err.message))
+const app = express()
 
 // Added Builtin Middlewares Here
+if (process.env.NODE_ENV === 'development') {
+    app.use(morgan('dev'))
```
Comment (Shazil2154) [inline index.js]: if we are implementing morgan we should use it for production as well

File: index.js [hunk 11]
```diff
@@ -1,28 +1,37 @@
...
 
+// Routes
 app.use('/api/patients', patientRoutes)
 app.use('/api/appointments', appointmentRoutes)
 
 app.get('/', (req, res) => {
     res.send('App is running...')
 })
 
+// Builtin Middlewares
```
Comment (Shazil2154) [inline index.js]: these are custom middlewares IMO

File: package.json [hunk 12]
```diff
@@ -4,14 +4,18 @@
...
   "license": "ISC",
   "dependencies": {
     "@joi/date": "^2.1.0",
+    "colors": "^1.4.0",
+    "dotenv": "^16.0.3",
     "express": "^4.18.1",
+    "express-async-handler": "^1.2.0",
     "joi": "^13.1.0",
     "joi-objectid": "^4.0.2",
     "mongoose": "^6.6.4",
```
Comment (Shazil2154) [inline package.json:21]: nodemon should be a dev dependency

File: utils/utils.js [hunk 13]
```diff
@@ -77,10 +77,57 @@ const getFee = (appointments, requiredCurrency) => {
     return fee
 }
 
+// will return most Popular pet
```
Comment (Shazil2154) [inline utils/utils.js]: function names explains its purpose. no need for the comments

File: utils/utils.js [hunk 14]
```diff
@@ -77,10 +77,57 @@ const getFee = (appointments, requiredCurrency) => {
...
+}
+
+// will return every pet Details
+const getEveryPetDetails = (patients) => {
+    const petsDetail = patients.reduce((acc, patient) => {
+        const petDetail = {};
+        petDetail.id = patient._id;
+        petDetail.name = patient.name;
+        petDetail.totalFeePaid = patient.appointments.reduce((sum, appointment) => {
+          if (appointment.isPaid) {
```
Comment (Shazil2154) [inline utils/utils.js]: you can create a function to perform both actions

**Review by Shazil2154 — CHANGES_REQUESTED** [review]

File: controllers/patientController.js (hunk 6 above)
Comment (ismaildev7) [reply controllers/patientController.js]: done

File: utils/utils.js (hunk 14 above)
Comment (ismaildev7) [reply utils/utils.js]: done

**Pull request merged**